### DISCUSSION TITLE
Ist NumLock möglich?

### DISCUSSION THREAD

**Discussion by dannys9** [Rojetto/ReNeo#77] [Q&A] [answered]
Ich versuche, mit ReNeo die NumLock-Taste zu „reaktivieren“. Ich habe in der layouts.json folgenden Eintrag (vorsichtshalber) für alle 6 Ebenen vorgenommen:

`"45+":[{"keysym": "Num_Lock", "vk": "VK_NUMLOCK", "label": "\u21ed"}, …],`

Auf der Bildschirmtastatur wird das geänderte Label auch angezeigt, also sollte ReNeo die Änderung grundsätzlich akzeptiert haben. Einen Tab gibt sie zwar nicht mehr aus, allerdings macht sie jetzt gar nichts mehr – weder als Standalone, noch im Erweiterungsmodus, autoNumlock an/aus ändert auch nichts.

Gibt es eine Möglichkeit, Numlock wieder zu verwenden? Mir geht’s dabei um die Maustasten-Funktion von Windows, die ich bei abgeschaltetem Numlock-Modus verwenden möchte bzw. im eingeschalteten Modus das Numpad.

**Comment by Rojetto** [comment]
Da funktioniert tatsächlich noch etwas nicht perfekt. Numlock und die Numlock-Taste sind **komisch**, insbesondere wenn man sie remappen will. Dewegen gibt es bereits einige Workarounds im Code, einer davon verhindert genau deinen Anwendungsfall. Für den nächsten Release kann ich mir nochmal anschauen, ob man das hingebogen bekommt.

Bis dahin kannst du aber folgendes probieren: `autoNumlock` auf `false` ist schon mal genau richtig. Dann kannst du den `VK_NUMLOCK` wie in deinem Schnippsel auf irgendeine Taste **außer** die echte Numlock-Taste legen (hoffentlich ist noch was übrig, z. B. Pause mit Scancode `"45"`). Mit den Maustasten-Funktionen kenne ich mich nicht aus, eventuell musst du dafür die restlichen Numpad-Mappings entfernen oder anpassen und verlierst damit die höheren Ebenen. Erfahrungen gerne berichten :)

**Comment by dannys9** [comment]
Danke für die schnelle Antwort. An sowas hatte ich auch schon gedacht. Ich hab jetzt mal wie vorgeschlagen VK_NUMLOCK auf Pause gelegt und bekomme damit folgendes Resultat:

Erweiterungsmodus: Taste reagiert nicht
Standalone: NumLock-LED schaltet um, der tatsächliche Numpad-Status bleibt jedoch bestehen. Drücke ich nach der Pause-Taste die eigentliche Numlock, wird das Numpad entsprechend der LED umgeschaltet. Danach muss ich dann [Pause][Pause][Numlock] drücken, um den Status erneut umzuschalten. Der Status wird nun auch von der Tastaturmaus-Funktion erkannt.

Hab’s mal mit T1 statt Pause probiert, gleiches Ergebnis.

Noch etwas ist mir aufgefallen. Die Tastaturmaus gibt beim Umschalten verschieden hohe Pieptöne zur Bestätigung aus. Egal, was ich auf die Numlock-Taste lege (also auch Tab), der Signalton wird immer ausgegeben. Ich vermute, dass Microsoft da selbst einen Hook verwendet, den du „von außen“ wohl nicht umgehen können wirst.

So wie oben funktioniert’s erstmal, besser als ReNeo immer zu (de-)aktivieren. Vielen Dank für den Tipp!

Vielleicht findest du ja noch eine bessere Lösung. Wäre es z. B. möglich, über eine Option in der config.json die Taste in ReNeo komplett zu ignorieren und stattdessen vom Tastaturtreiber verarbeiten zu lassen? Wenn das ginge, dann sollte die Numlock auch im Erweiterungsmodus wieder funktionieren.

**Comment by Rojetto** [comment, marked as answer]
In [v1.5.4](https://github.com/Rojetto/ReNeo/releases/tag/v1.5.4) kannst du jetzt einfach die Mappingzeile für `"45+"` löschen, dann wird Numlock unangetastet durchgelassen. Bei mir funktioniert es so sowohl im Standalone- als auch im Erweiterungsmodus.

**Comment by dannys9** [comment]
Whoah, das ging fix 😁
Funktioniert, dankeschön!
